### PR TITLE
Isthmus: L2 Withdrawals root spec updates

### LINKED ISSUE #394: Isthumus: Spec updates for L2 withdrawals-root in block header
Peeling off the work associated with updating specs mentioned in: https://github.com/ethereum-optimism/optimism/issues/12044

General notes: keep the specs categorized in the `isthumus ` directory, and file names consistent across upgrades:
e.g. Execution layer changes go `exec-engine.md`.

- [x] Specs note that the block-body withdrawals list is encoded as an empty RLP list.
- [x] Specs note that any block-after-transaction simulation should include an empty withdrawals-root;
  the block is sealed at the end of a block.
- [x] Specs note that the withdrawals-root in the genesis block,
  if Isthumus is active, is always the empty withdrawals root, regardless of L2 state.
- [x] Specs note that the withdrawals-root starting at shanghai, pre-Isthumus,
  is set to an empty withdrawals list, this equals the same root as an empty storage root.
  The withdrawals are there in L2 state, just not elevated into the withdrawals-root.
  Pre-Isthumus output-root construction should be careful not to use the header withdrawals-root.
- [x] Specs note that at the time of state-processing, with an in-progress header,
  the withdrawals-root should not be available to the EVM / application layer.
- [x] Specs Engine-API modification: `ExecutableData` gets a `withdrawalsRoot` attribute. 
- [x] Specs note that for an accurate storage-root to be known,
  the block state has to be committed first, before the withdrawals storage root can be set accurately.
  This attribute should thus be set right after setting the state-root attribute of the header.
- [x] Specs EL P2P note: block-bodies and block-headerts responses are handled the same in isolation:
  we compute the body-withdrawal-hashes from the received withdrawals list in the block body.
  Then, when we compare the body-withdrawal-hashes to the header withdrawal-hashes,
  and if optimism Isthumus is active for the given header, we check if the body-withdrawal-hash matches the empty list root,
  while allowing the header-withdrawal-hash to be any non-null value.
  By verifying at this later stage, we have the header timestamp, and can apply the fork logic conditionally.
- [x] Specs update to describe CL `ExecutionPayload` SSZ encoding with new withdrawals-root included.
    - ~[ ] We remove the `withdrawals` list from the encoding, diverging from the L1 beacon-chain `ExecutionPayload` encoding~.
- [x] Spec that we continue to use `engine_newPayloadV3` engine-API,
  but now with the additional `ExecutionPayload` attribute, that is omitted pre-Isthumus.
- [x] Specs update to describe new CL P2P gossip topic, with updated `ExecutionPayload` SSZ encoding,
  and a gossip-validation rule that checks if the withdrawals root is non-nil, and withdrawals list is empty.
  This spec change goes into the main `rollup-node-p2p.md`, to be consistent with the other gossip topics.
- [x] Update fault-proof-program spec that describes the output-claim verification,
  and link to the Isthumus specs change of the block header.

### PR DESCRIPTION
**Pull request by vdamle**
**Description**

Spec Updates for Isthumus: L2 Withdrawals root spec updates

**Tests**

None

**Metadata**

- Fixes #394

### CODE REVIEW

File: specs/protocol/holocene/exec-engine.md
Comment (tynes) [inline specs/protocol/holocene/exec-engine.md]: A good way to make this very clear is saying that its the storage root corresponding to the post state, exactly how the post global state root is included in the block header

File: specs/protocol/holocene/exec-engine.md
Comment (tynes) [inline specs/protocol/holocene/exec-engine.md]: Should be careful to only use the withdrawals root after holocene is active? which can be determined if its not empty

**Comment by tynes** [comment]
Nice work on this so far!

File: specs/protocol/holocene/exec-engine.md
Comment (protolambda) [inline specs/protocol/holocene/exec-engine.md]: at the genesis block

missing "the"

File: specs/protocol/holocene/exec-engine.md
Comment (protolambda) [inline specs/protocol/holocene/exec-engine.md]: not set to an empty list, but set to the MPT root of an empty list

File: specs/protocol/holocene/exec-engine.md
Comment (protolambda) [reply specs/protocol/holocene/exec-engine.md]: If the smart contract storage is empty, it matches the empty withdrawals list, even though it represents something different pre-holocene.

And pre-holocene, even though the withdrawals root is present, and a valid MPT root in the block header, it shouldn't be used.

And post-holocene, if there was never any withdrawal contract storage, a MPT root equal to one of an empty list, can be valid to use.

Not sure how to best word a warning for these possible easy things to mistake, but I think it's worth mentioning something to this extent.

**Review by protolambda — COMMENTED** [review]
Great start

File: specs/protocol/holocene/exec-engine.md
Comment (vdamle) [reply specs/protocol/holocene/exec-engine.md]: added this on at the end of the current description.

File: specs/protocol/rollup-node-p2p.md
Comment (vdamle) [inline specs/protocol/rollup-node-p2p.md:287]: Question: Is my assumption correct that this is an additive list?

File: specs/protocol/rollup-node-p2p.md
Comment (vdamle) [inline specs/protocol/rollup-node-p2p.md:419]: From the proposed changes [here](https://github.com/ethereum-optimism/op-geth/pull/383/files#diff-50e0dca0bd0ac667ca6a4a3b762c9346030f346fdd9b48f3f83be675a43423eaR91-R94) and as further elaborated [here](https://github.com/ethereum-optimism/optimism/issues/12044#issuecomment-2369618003), the `ExecutionPayload` (and not `ExecutionPayloadEnvelope`) will be extended with `withdrawalsRoot`

File: specs/protocol/rollup-node-p2p.md
Comment (protolambda) [reply specs/protocol/rollup-node-p2p.md:287]: the block header RLP list? Kind of, attributes have been reused / changed meaning in the past, but it's meant to be backward compatible.
The SSZ version that is used for Gossip, with the `parentBeaconBlockRoot` added to it, is already non-standard, and not strictly speaking SSZ anymore (due to a missing offset value).

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: I'm not sure what this actually means.  Is this when the L2 block's origin is an L1 block with timestamp > activation or if the L2 block came from a batch included on L1 after activation time (and is that when the channel starts or ends?).

Also, why would different parts of the L2 block activate Isthmus at different times?

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: Would it be simpler to say this should be the account store root from the world state identified by the stateRoot field in the block header?

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: This seems like an implementation detail which isn't necessarily true. The withdrawals root could easily be set once its final state is known even if hashing some other part of the world state trie is ongoing to calculate the overall state root.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: Why wouldn't this be set to match the actual L2 state? Given genesis is the initial safe head having this be inconsistent could lead to creating invalid proposals.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: Do we need to be clear about what this validation is and exactly when it is performed? ie does a node have to go back and recheck the header of every block after sync completes or just performs that check as each header is received?

Actually, how does a node do that verification if it's snap syncing given it won't have the world state? I _think_ just verifying the final block is enough since it's parent hash commits to the previous world states and the withdrawalsRoot header and snap sync assumes that the chain it downloads has valid execution of transactions (since it skips executing them).

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: Why is this empty and not the actual root?  Is the state root field not updated for the simulated transactions?

Also, what does simulated transactions mean in this context? I'm assuming for some RPC calls but we probably should be clear about the context here.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]:
```suggestion
Withdrawals list in the block body is encoded as an empty RLP list.
```

**Review by ajsutton — COMMENTED** [review]
Looks good generally. I've left a bunch of comments but primarily aimed at ensuring we're being really clear in the spec rather than disagreeing with the design (apart from the withdrawalRoot for genesis which seems weird to me but maybe I'm missing something).

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [reply specs/protocol/isthmus/exec-engine.md]: This is true, if the world state is being built concurrently, so its an implementation detail. The most important thing is that its the post state

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [reply specs/protocol/isthmus/exec-engine.md]: This saves some effort when building the L2 genesis. If its going to break the proposer, its definitely something that we can support. There is not technically much utility about making a claim about the genesis state to facilitate a withdrawal since there are none in the genesis state of a standard config chain

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [reply specs/protocol/isthmus/exec-engine.md]: Verifying the final block is indeed enough, but this is definitely a weird edge case around reorging that I am not exactly sure how snapsync handles

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [inline specs/protocol/isthmus/exec-engine.md]: Which OP Stack is shanghai in? Could be better to use the OP Stack fork name

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [reply specs/protocol/isthmus/exec-engine.md]: The withdrawals root isn't in EVM context? Transaction simulation via `eth_call` doesn't build a block?

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: I believe it is the former -- the L2 block's origin is an L1 block with timestamp > activation.

> Also, why would different parts of the L2 block activate Isthmus at different times?

Definitely didn't mean to imply the above. Also not sure where I've implied the above! Please let me know and I'll fix it.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Thanks for the suggestion, that definitely sounds more straight forward. Will make the change

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Ah, makes sense! I've removed this.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [reply specs/protocol/isthmus/exec-engine.md]: It's not so much that there's utility in proposing for the genesis block, but that this breaks two invariants:
1. That starting from Isthmus the withdrawalsRoot header is the `L2ToL1MessagePasser` storage root
2. That the safe head can always be used to produce a valid proposal

Everywhere that reads the withdrawalRoot would need to have special case handling for genesis. So I'd be very inclined to have the extra state read in L2 genesis generation to make all other cases simple.

Comment (ajsutton) [reply specs/protocol/isthmus/exec-engine.md]: As an example of the bugs that come from this special case - this PR adds a paragraph to `proposal.md` that says the withdrawalsRoot can be used post-Isthmus but forgets the genesis special case where it can't.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Good point - Per [this issue](https://github.com/ethereum-optimism/optimism/issues/6728) and [blog post](https://blog.oplabs.co/canyon-hardfork/) , it is in `Canyon` hard fork. Modified to specify Canyon.

File: specs/protocol/isthmus/exec-engine.md
Comment (protolambda) [reply specs/protocol/isthmus/exec-engine.md]: The withdrawal-contract genesis storage-root will keep changing, and having many special genesis values to put in the block-header makes things complicated. It couples the state content and block-header generation to tightly. We should just default the withdrawals-root in the genesis block to an empty withdrawals MPT root. I think not being allowed to withdraw from the genesis block is fine, as there aren't supposed to be withdrawals anyway.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Agree the wording makes it unclear, will clarify that the validation is performed on the final synced block.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: > As an example of the bugs that come from this special case - this PR adds a paragraph to proposal.md that says the withdrawalsRoot can be used post-Isthmus but forgets the genesis special case where it can't.

🤦🏼‍♂️ Indeed, I will add that.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: I see that the wording doesn't make clear the situations in which this is applicable. There is a [`eth_simulate`](https://github.com/ethereum/execution-apis/pull/484) proposed addition where transactions can be simulated spread across multiple blocks. The intent here is to specify that in such cases, an empty withdrawal root is included in the block header. Will clarify this.

File: specs/protocol/isthmus/exec-engine.md
Comment (tynes) [inline specs/protocol/isthmus/exec-engine.md]: nit: Isthmus is capitalized elsewhere, we should be standard

File: specs/protocol/rollup-node-p2p.md
Comment (tynes) [inline specs/protocol/rollup-node-p2p.md:271]: I do not think we need to add a new p2p topic for holocene anymore, so this should be good, waiting on a response: https://github.com/ethereum-optimism/specs/pull/428#issuecomment-2418099412

**Review by tynes — APPROVED** [review]
This looks good to me, would like approval from @ajsutton or @protolambda before merge

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [reply specs/protocol/isthmus/exec-engine.md]: Typically to create genesis block you populate the world state from the genesis config and then insert that into the stateRoot field of the block header. That already includes the withdrawal contract storage root so there doesn't seem to be any additional dependencies being added here.

My concern is less about whether you can withdraw but more that it's really easy to forget genesis is a special case and do things like make invalid proposals or depend on the withdrawalsRoot as mentioned above when it's not actually accurate. Given we already have to know the genesis world state to create the block header it doesn't seem worth creating that foot gun.

File: specs/protocol/rollup-node-p2p.md
Comment (vdamle) [reply specs/protocol/rollup-node-p2p.md:271]: Thanks for flagging that! Correct, @roberto-bayardo responded to your clarification - https://github.com/ethereum-optimism/specs/pull/428#issuecomment-2418259114

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [reply specs/protocol/isthmus/exec-engine.md]: I think what I'm struggling with is this paragraph both says changes to the L2 block execution rules activate when L2 timestamp >= activation time (existed prior to this PR) _and_ the block header changes apply when considering data from a L1 block with timestamp >= activation time (added in this PR). So the L2 block changes are activating at different times which doesn't make sense to me.

For L2 block changes, I'd expect the first condition to be right - the changes activate when the L2 block timestamp >= activation time and it doesn't matter when that batch data is actually included on L1 or what the L1 origin timestamp is.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: I'd just say:

```
After Isthmus hardfork's activation, the L2 block header's `withdrawalsRoot` field will consist of the 32-byte
 [`L2ToL1MessagePasser`][l2-to-l1-mp] account storage root from the world state identified but the stateRoot field in the block header.
```

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md]: I've got to admit, I still don't understand this. There are no withdrawals in any block for OP Stack so wouldn't the computed withdrawal root always be the empty hash? And when the final header is received if Isthmus is active shouldn't it have the message passer account storage root?

I know headers and bodies in devp2p are matched up by comparing roots (ie tx root, receipts root and withdrawals root) so during sync if you request a batch of block bodies, you'd expect that they all have an empty withdrawals list since OP Stack doesn't use withdrawals.  To match them up with the headers you'd have to essentially ignore the withdrawalsRoot (or treat the header as if it had an empty withdrawal root). I think this is trying to explain that.

File: specs/protocol/isthmus/exec-engine.md
Comment (ajsutton) [inline specs/protocol/isthmus/exec-engine.md:108]: unless the block is genesis (though I think we should fix that).

File: specs/protocol/proposals.md
Comment (ajsutton) [inline specs/protocol/proposals.md]: This leaves some ambiguity about what the correct output root should be.  Is the withdrawals root taken from the genesis header or from the genesis world state when constructing the output root for the genesis block?

File: specs/protocol/rollup-node-p2p.md
Comment (ajsutton) [reply specs/protocol/rollup-node-p2p.md:271]: Note that at the top of the Gossip Topics section it says "There are three topics" but now there are four (we should rephrase that to not specify a number).

**Review by ajsutton — COMMENTED** [review]
Looks good generally, just a couple of bits that I think could use more clarity.  For the P2P stuff I don't have all the context so may just be misunderstanding things.

I am very strongly of the opinion that it's a mistake to have the genesis root hard code an empty hash for withdrawalsRoot instead of using the actual value from the account storage. The `stateRoot` field will be changing with any changes to the actual storage root anyway.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Removed the confusing new statement. The original description of timestamp activation conveys the intended behavior.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: I met with @protolambda / @ajsutton and we came to the agreement that even for the genesis block, we will set the `withdrawalsRoot` field in the header to the `L2ToL1MessagePasser` storage root, if Isthmus is active -- this is a change from the original spec proposal that stated that it will be set to the empty withdrawals root.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md:108]: no longer relevant as genesis header withdrawalsRoot shall also be an MPT root of an empty list (we decided not to special case the genesis withdrawalsRoot).

File: specs/protocol/proposals.md
Comment (vdamle) [reply specs/protocol/proposals.md]: Updated.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: I've updated the description based on the conversation yesterday.

File: specs/protocol/isthmus/exec-engine.md
Comment (vdamle) [reply specs/protocol/isthmus/exec-engine.md]: Updated based on discussion yesterday.

File: specs/protocol/holocene/derivation.md
Comment (vdamle) [inline specs/protocol/holocene/derivation.md:138]: fixed an incorrect link - this was not meant to reference a url declaration but just another section in the doc.

**Review by ajsutton — APPROVED** [review]
LGTM.

**Review by protolambda — APPROVED** [review]

**Pull request merged**
